{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"axis1.ipynb","provenance":[],"collapsed_sections":[],"authorship_tag":"ABX9TyOBuy7OBtks1yoiDUb5o/R0"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":null,"metadata":{"id":"NV5WpdBLKsUq"},"outputs":[],"source":["%load_ext autoreload\n","%autoreload 2"]},{"cell_type":"code","source":["import sys\n","import pickle as pkl\n","\n","from sklearn.metrics import precision_recall_fscore_support\n","\n","import string\n","import logging\n","\n","import glob\n","import tqdm\n","import pickle\n","\n","from nlstruct.utils import torch_clone\n","from nlstruct.utils import torch_global as tg\n","from nlstruct.dataloaders import load_from_brat\n","\n","\n","sys.path.insert(0,'/export/home/cse200093/deep_mlg_normalization/deep_multilingual_normalization')\n","sys.path.insert(0,'/export/home/cse200093/deep_mlg_normalization')\n","\n","from deep_multilingual_normalization.preprocess import preprocess, load_quaero\n","from deep_multilingual_normalization.train import train_step1, train_step2, clear\n","from deep_multilingual_normalization.eval import predict\n","\n","from notebook_utils import *\n","\n","print('done')"],"metadata":{"id":"5X5R7I2vKx7z"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# load cui2sty\n","open_file = open('/export/home/cse200093/Expe_Translation/english_Norm/umls_en_fr_cui2sty.pkl', \"rb\")\n","umls_cui2sty = pickle.load(open_file)\n","open_file.close()"],"metadata":{"id":"Sz85Q1s4KzXj"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# map sty group to semantic group\n","sty2type = {\n","    'Activity': 'ACTI',\n","    'Behavior': 'ACTI',\n","    'Daily or Recreational Activity': 'ACTI',\n","    'Event': 'ACTI',\n","    'Governmental or Regulatory Activity': 'ACTI',\n","    'Individual Behavior': 'ACTI',\n","    'Machine Activity': 'ACTI',\n","    'Occupational Activity': 'ACTI',\n","    'Social Behavior': 'ACTI',\n","    'Anatomical Structure': 'ANAT',\n","    'Body Location or Region': 'ANAT',\n","    'Body Part, Organ, or Organ Component': 'ANAT',\n","    'Body Space or Junction': 'ANAT',\n","    'Body Substance': 'ANAT',\n","    'Body System': 'ANAT',\n","    'Cell': 'ANAT',\n","    'Cell Component': 'ANAT',\n","    'Embryonic Structure': 'ANAT',\n","    'Fully Formed Anatomical Structure': 'ANAT',\n","    'Tissue': 'ANAT',\n","    'Amino Acid, Peptide, or Protein': 'CHEM',\n","    'Antibiotic': 'CHEM',\n","    'Biologically Active Substance': 'CHEM',\n","    'Biomedical or Dental Material': 'CHEM',\n","    'Chemical': 'CHEM',\n","    'Chemical Viewed Functionally': 'CHEM',\n","    'Chemical Viewed Structurally': 'CHEM',\n","    'Clinical Drug': 'CHEM',\n","    'Element, Ion, or Isotope': 'CHEM',\n","    'Enzyme': 'CHEM',\n","    'Hazardous or Poisonous Substance': 'CHEM',\n","    'Hormone': 'CHEM',\n","    'Immunologic Factor': 'CHEM',\n","    'Indicator, Reagent, or Diagnostic Aid': 'CHEM',\n","    'Inorganic Chemical': 'CHEM',\n","    'Nucleic Acid, Nucleoside, or Nucleotide': 'CHEM',\n","    'Organic Chemical': 'CHEM',\n","    'Pharmacologic Substance': 'CHEM',\n","    'Receptor': 'CHEM',\n","    'Vitamin': 'CHEM',\n","    'Classification': 'CONC',\n","    'Conceptual Entity': 'CONC',\n","    'Functional Concept': 'CONC',\n","    'Group Attribute': 'CONC',\n","    'Idea or Concept': 'CONC',\n","    'Intellectual Product': 'CONC',\n","    'Language': 'CONC',\n","    'Qualitative Concept': 'CONC',\n","    'Quantitative Concept': 'CONC',\n","    'Regulation or Law': 'CONC',\n","    'Spatial Concept': 'CONC',\n","    'Temporal Concept': 'CONC',\n","    'Drug Delivery Device': 'DEVI',\n","    'Medical Device': 'DEVI',\n","    'Research Device': 'DEVI',\n","    'Acquired Abnormality': 'DISO',\n","    'Anatomical Abnormality': 'DISO',\n","    'Cell or Molecular Dysfunction': 'DISO',\n","    'Congenital Abnormality': 'DISO',\n","    'Disease or Syndrome': 'DISO',\n","    'Experimental Model of Disease': 'DISO',\n","    'Finding': 'DISO',\n","    'Injury or Poisoning': 'DISO',\n","    'Mental or Behavioral Dysfunction': 'DISO',\n","    'Neoplastic Process': 'DISO',\n","    'Pathologic Function': 'DISO',\n","    'Sign or Symptom': 'DISO',\n","    'Amino Acid Sequence': 'GENE',\n","    'Carbohydrate Sequence': 'GENE',\n","    'Gene or Genome': 'GENE',\n","    'Molecular Sequence': 'GENE',\n","    'Nucleotide Sequence': 'GENE',\n","    'Geographic Area': 'GEOG',\n","    'Age Group': 'LIVB',\n","    'Amphibian': 'LIVB',\n","    'Animal': 'LIVB',\n","    'Archaeon': 'LIVB',\n","    'Bacterium': 'LIVB',\n","    'Bird': 'LIVB',\n","    'Eukaryote': 'LIVB',\n","    'Family Group': 'LIVB',\n","    'Fish': 'LIVB',\n","    'Fungus': 'LIVB',\n","    'Group': 'LIVB',\n","    'Human': 'LIVB',\n","    'Mammal': 'LIVB',\n","    'Organism': 'LIVB',\n","    'Patient or Disabled Group': 'LIVB',\n","    'Plant': 'LIVB',\n","    'Population Group': 'LIVB',\n","    'Professional or Occupational Group': 'LIVB',\n","    'Reptile': 'LIVB',\n","    'Vertebrate': 'LIVB',\n","    'Virus': 'LIVB',\n","    'Entity': 'OBJC',\n","    'Food': 'OBJC',\n","    'Manufactured Object': 'OBJC',\n","    'Physical Object': 'OBJC',\n","    'Substance': 'OBJC',\n","    'Biomedical Occupation or Discipline': 'OCCU',\n","    'Occupation or Discipline': 'OCCU',\n","    'Health Care Related Organization': 'ORGA',\n","    'Organization': 'ORGA',\n","    'Professional Society': 'ORGA',\n","    'Self-help or Relief Organization': 'ORGA',\n","    'Biologic Function': 'PHEN',\n","    'Environmental Effect of Humans': 'PHEN',\n","    'Human-caused Phenomenon or Process': 'PHEN',\n","    'Laboratory or Test Result': 'PHEN',\n","    'Natural Phenomenon or Process': 'PHEN',\n","    'Phenomenon or Process': 'PHEN',\n","    'Cell Function': 'PHYS',\n","    'Clinical Attribute': 'PHYS',\n","    'Genetic Function': 'PHYS',\n","    'Mental Process': 'PHYS',\n","    'Molecular Function': 'PHYS',\n","    'Organism Attribute': 'PHYS',\n","    'Organism Function': 'PHYS',\n","    'Organ or Tissue Function': 'PHYS',\n","    'Physiologic Function': 'PHYS',\n","    'Diagnostic Procedure': 'PROC',\n","    'Educational Activity': 'PROC',\n","    'Health Care Activity': 'PROC',\n","    'Laboratory Procedure': 'PROC',\n","    'Molecular Biology Research Technique': 'PROC',\n","    'Research Activity': 'PROC',\n","    'Therapeutic or Preventive Procedure': 'PROC'\n","}"],"metadata":{"id":"fQxZQ4c4LAmD"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# path to predict data and gold standard\n","# the NER phase has been done in nlstruct_predict.ipynb and results saved to PREDICT_DATA_PATH\n","PREDICT_DATA_PATH = '/export/home/cse200093/Expe_Translation/nlstruct_EMEA_restrict_types'\n","STAND_DATA_PATH = '/export/home/cse200093/brat_data/QUAERO_FrenchMed/corpus/test/EMEA'"],"metadata":{"id":"Wcfdq3uXLGga"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# load the normalization model\n","model = torch.load('/export/home/cse200093/deep_mlg_normalization/models/UMLS2021AB_without_quaero.pt')\n","model.eval()"],"metadata":{"id":"aQ4ekk5gLMn7"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["bert_name = \"/export/home/cse200093/deep_mlg_normalization/bert-base-multilingual-uncased\"\n","VOCAB_PATH = '/export/home/cse200093/deep_mlg_normalization'\n","\n","import os\n","\n","with open(os.path.join(VOCAB_PATH,'vocab1.pkl'), 'rb') as f:\n","    vocabularies1 = pkl.load(f)\n","\n","with open(os.path.join(VOCAB_PATH,'vocab2021AB_v2.pkl'), 'rb') as f:\n","    vocabularies2 = pkl.load(f)"],"metadata":{"id":"W7mg0X2JLuP1"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["dataset = load_from_brat(PREDICT_DATA_PATH)\n","\n","dataset['mentions']['mention_id'] = dataset['mentions']['doc_id'] +'.'+ dataset['mentions']['mention_id'].astype(str)\n","\n","batcher, vocs, mention_ids = preprocess_train(\n","    dataset,\n","    vocabularies=vocabularies2,\n","    bert_name=bert_name,\n",")\n","\n","batch_size = len(batcher)\n","with_tqdm = True\n","\n","tg.set_device('cuda:6') #('cuda:0')\n","device = tg.device\n","\n","# topk: return first k candidates\n","nb = 5\n","pred_batcher = predict(batcher, model, batch_size=64, return_loss=False, topk=nb)\n","\n","print('pred_batcher', pred_batcher)\n","print('batcher', batcher)"],"metadata":{"id":"m_vxDRtuLwl6"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["voc_tokens = vocs['token']\n","voc_labels = vocs['label']\n","def replace_fn(s):\n","    return s.replace(' ##', '').replace('[SEP]', '').replace('[PAD]', '').replace('[CLS]', '').strip()\n","merged_batcher = batcher['mention',['mention_id','token']].merge(pred_batcher)\n","final_tokens = [replace_fn(' '.join([voc_tokens[i] for i in b])) for b in merged_batcher['mention']['token'].toarray()]\n","final_labels = [voc_labels[b] for b in merged_batcher['mention']['label']]\n","\n","final_couples = [(t, l) for t,l in zip(final_tokens, final_labels)]\n","\n","# index of elements in the list is just the corresponding mention_id\n","final_couples"],"metadata":{"id":"nMrhvoXqLyCS"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Add type information: output the first candidate with right type\n","def choose(l, stand):\n","    i = 0\n","    while l[i]!=stand and i<nb-1:\n","        i+=1\n","    if i == nb-1:\n","        return 0\n","    else:\n","        return i\n","    \n","def type_info(final_couples):\n","    gold_types = list(dataset['mentions']['label'])\n","    new_final_couples = []\n","    nb_unknown = 0 # number of unknown cuis\n","    unknown_cuis = []\n","    for i in tqdm.tqdm(range(len(final_couples))):\n","        candidates = final_couples[i][1]\n","        candidate_types = [sty2type[umls_cui2sty[candidate]] if candidate in umls_cui2sty.keys() else 'unknown_cui' for candidate in candidates]\n","        nb_unknown+=len([x for x in candidate_types if x=='unknown_cui'])\n","        unknown_cuis+=[candidates[i] for i in range(nb) if candidate_types[i]=='unknown_cui']\n","        gold_type = gold_types[i]\n","        res = choose(candidate_types, gold_type)\n","        \n","        new_final_couples.append((final_couples[i][0],candidates[res]))\n","        \n","    print(nb_unknown)\n","    return (new_final_couples,unknown_cuis)"],"metadata":{"id":"--nSPgOTLzlU"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# add type info\n","final_couples,unknown_cuis = type_info(final_couples)\n","final_couples"],"metadata":{"id":"rO8CdGHqL087"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["df_mentions = dataset['mentions'].copy()\n","df_mentions['cui'] = [x[1] for x in final_couples]\n","df_mentions.set_index('mention_id',inplace=True)\n","df_mentions"],"metadata":{"id":"kF8UXfitL2us"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["os.chdir(STAND_DATA_PATH)\n","my_files = glob.glob('*.ann')\n","len(my_files)"],"metadata":{"id":"gbkIwEBiL4F6"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["phrases = []\n","cuis = []\n","types = []\n","dic = {} # necessary when len(phrases) != len(cuis)\n","for file in tqdm.tqdm(my_files):\n","    phrase = None\n","    code = None\n","    type_name = None\n","    f = open(STAND_DATA_PATH+'/'+file, \"r\")\n","    doc = file\n","    for line in f:\n","        line = line.rstrip('\\n')\n","        if line.startswith('T'):\n","            type_name = line.split('\\t')[1].split(' ')[0]\n","            phrase = line.split('\\t')[2]\n","            code = line.split('\\t')[0] # necessary when len(phrases) != len(cuis)\n","            dic[file+'.'+phrase+'.'+code] = 'no cui' # necessary when len(phrases) != len(cuis)\n","            #phrases.append(phrase)\n","            #types.append(type_name)\n","        elif line.startswith('#'):\n","            cui = line.split('\\t')[2]\n","            dic[file+'.'+phrase+'.'+code] = (cui,type_name) # necessary when len(phrases) != len(cuis)\n","            #cui = cui.split(',')[0][1:].rstrip('\\\"') # necessary for Mantra\n","            #cuis.append(cui)"],"metadata":{"id":"AcIOH7eTL5rr"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# necessary when len(phrases) != len(cuis)\n","print(len(dic))\n","no_cuis = []\n","docs = []\n","# phrases_eng = []\n","for key in tqdm.tqdm(dic):\n","    if dic[key] != 'no cui':\n","        phrases.append(key.split('.')[2])\n","        # necessary with translation\n","        # phrases_eng.append(translator.translate(key.split('.')[2]).text)\n","        cuis.append(dic[key][0])\n","        types.append(dic[key][1])\n","        docs.append(key.split('.')[0])\n","    else:\n","        #print(key)\n","        no_cuis.append(key)\n","        \n","print(len(cuis),len(phrases),len(types))"],"metadata":{"id":"wpB6asjqL7iT"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# gold standard\n","df_stand = pd.DataFrame({'doc_id':docs, 'text':phrases, 'cui':cuis, 'types':types})\n","df_stand = df_stand[(df_stand['types']=='PROC')|(df_stand['types']=='DEVI')|(df_stand['types']=='DISO')|(df_stand['types']=='CHEM')]\n","df_stand"],"metadata":{"id":"LEDeL7BkL8tM"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["def get_dic(df):\n","    dic = {}\n","    docs = list(df['doc_id'])\n","    for doc in docs:\n","        cuis = list(df[df['doc_id']==doc]['cui'])\n","        dic[doc] = cuis\n","    return dic"],"metadata":{"id":"5gxMD587MAAC"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["def convert_list_to_dic(dic):\n","    for file_name in tqdm.tqdm(dic):\n","        key_list = list(set(dic[file_name]))\n","        value_list = [dic[file_name].count(x) for x in key_list]\n","        res = dict(zip(key_list, value_list))\n","        dic[file_name] = res"],"metadata":{"id":"mPHOb2CqMBkM"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["dic_stand = get_dic(df_stand)\n","dic_res = get_dic(df_mentions)\n","\n","convert_list_to_dic(dic_stand)\n","convert_list_to_dic(dic_res)"],"metadata":{"id":"RJrVVkuQMClR"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["df_compare = pd.DataFrame(columns = ['file_name','correct_ents','built_model_ents'])\n","\n","for file in tqdm.tqdm(dic_stand):\n","    df_compare = df_compare.append({'file_name':file, 'correct_ents':dic_stand[file], \n","                       'built_model_ents':dic_res[file]},\n","                     ignore_index=True)\n","    \n","df_compare"],"metadata":{"id":"YoB0-cxOMFMi"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["df_compare['built_TP'] = 0\n","df_compare['built_FP'] = 0\n","df_compare['built_FN'] = 0\n","\n","df_compare['built_precision'] = 0\n","df_compare['built_recall'] = 0\n","df_compare['built_f1-score'] = 0\n","\n","# calculate TP,FP,FN for each model\n","for index, row in tqdm.tqdm(df_compare.iterrows()):\n","    for cui in row['built_model_ents']:\n","        if cui in row['correct_ents']:\n","            df_compare.loc[index,'built_TP'] += min(row['correct_ents'][cui], row['built_model_ents'][cui])\n","        else:\n","            df_compare.loc[index,'built_FP'] += row['built_model_ents'][cui]\n","    for cui in row['correct_ents']:\n","        if cui not in row['built_model_ents']:\n","            df_compare.loc[index,'built_FN'] += row['correct_ents'][cui]\n","            \n","            \n","# calculate precision and recall for each model\n","for index, row in tqdm.tqdm(df_compare.iterrows()):\n","    if row['built_TP'] != 0:\n","        df_compare.loc[index,'built_precision'] = row['built_TP']/(row['built_TP']+row['built_FP'])\n","\n","    if row['built_TP'] != 0:\n","        df_compare.loc[index,'built_recall'] = row['built_TP']/(row['built_TP']+row['built_FN'])\n","\n","# calculate f1-score for each model\n","for index, row in tqdm.tqdm(df_compare.iterrows()):\n","    if row['built_TP'] != 0:\n","        df_compare.loc[index,'built_f1-score'] = (2*row['built_precision']*row['built_recall'])/(row['built_precision']+row['built_recall'])\n","            \n","\n","df_compare"],"metadata":{"id":"szb-KHz2MGsb"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# calculate mean precision, mean recall, mean f1-score\n","\n","built_avg_precision = np.mean(df_compare['built_precision'])\n","built_avg_recall = np.mean(df_compare['built_recall'])\n","built_avg_f1_score = np.mean(df_compare['built_f1-score'])\n","\n","print(f'Average precision of the built model: {built_avg_precision}')\n","print(f'Average recall of the built model: {built_avg_recall}')\n","print(f'Average f1-score of the built model: {built_avg_f1_score}')"],"metadata":{"id":"2bhsyQI1MIzK"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# save df_compare for boostrap\n","df_compare.to_pickle('/export/home/cse200093/Expe_Translation/Axes_script/EMEA_res.pkl')"],"metadata":{"id":"lf_aw35AMKXC"},"execution_count":null,"outputs":[]}]}